### ISSUE TITLE
Wrong Password?

### ISSUE THREAD

**Issue by wrldtvlr** [josephdadams/TallyArbiter#401]
Just attempted a CLI update to 3.0.2 on a Raspberry Pi 4. Update seemed to have gone fine, auto restart works, but when I attempt to login, I get a wrong password error. 

The config json file had a long character strings for both the settings and producer passwords. I assume they were encrypted, but I tried changing them to 12345 per current docs to no avail.

Suggestions?

**Comment by wrldtvlr** [comment]
We took a look at the code this morning to see if we could temporarily comment out the password check. While looking for that we noticed there has been a update to the code to use the bcrypt library - perhaps to store a hash of the password in config.json instead of the raw password. Didn't have enough time to dig further, but suspect this hasn't been fully tested.

**Comment by JTF4** [comment]
Hey @wrldtvlr! Would you mind uploading a copy of your config json?

We've been testing and using the new authentication system successfully for a while now with no issue.

**Comment by hrueger** [comment]
Hi @wrldtvlr,
`null` or `12345` won't work because the password from the config file must always be a hash. When logging in, the password from the input field is [compared with the hash](https://github.com/josephdadams/TallyArbiter/blob/9e4f6c00714832ccf2cbe43103a63920b38a2d28/src/_helpers/auth.ts#L22).
You could try something like https://bcrypt-generator.com/ to generate a suitable hash and put it in your config.

In the future, we could add like a reset password CLI option.

**Comment by wrldtvlr** [comment]
Still not clear what is going on here. Generated a hash for 12345 at bcrypt-generator.com, pasted it in, rebooted. Accessing the settings page still results in "Wrong username or password!".

**Comment by wrldtvlr** [comment]
Oh - How many rounds should I be using when generating the hash?

**Comment by wrldtvlr** [comment]
Looked at the code linked above and thought the number of rounds might be 10. This resulted in the hash below, which did not work.

   "username": "admin",
   "password": "$2a$10$5c81Ep72a3PUK8qTRI.19uvVEJhosg2QLrGkVPpOg6vwWxtSANlW2",
   "roles": "admin"

**Comment by MatteoGheza** [comment]
@wrldtvlr can you try running TA from the new branch fix_auth?
I'm sorry, I was very busy in these weeks, and I didn't tried the config migration a lot before merging...
There was a bug in the config migration system, so the JWT Private Key (required for JWT-based auth) was not generated, and the login function failed. Now every login error is logged into TA Logs.
It should work now.
I'm opening a PR right now.

**Comment by wrldtvlr** [comment]
I'll be in the booth tomorrow morning and give it a try!

**Comment by wrldtvlr** [comment]
Is the 10 round hash I posted above correct?

**Comment by MatteoGheza** [comment]
> Is the 10 round hash I posted above correct?

Yes, it is

**Comment by hrueger** [comment]
As far as I know, more rounds just make the hash more secure - and as the comparison function does not need to know how many rounds were used when generating the hash, I'd say that any number of rounds should work.

**Comment by MatteoGheza** [comment]
@hrueger can we detect if a string is a valid bcrypt-hashed string or just a string?
It would be awesome if TA detects plaintext password, hash them and save the config on startup.
Do you think it's possible?

**Comment by hrueger** [comment]
I think that the generated hashes are always 60 chars (or 59 according to stackoverflow). Maybe we can just check if the length is different and if yes, treat it as a plaintext password?

**Comment by wrldtvlr** [comment]
Sorry - my knowledge of npm is very limited. We're running TA using the CLI install process. I know how to update to a specific version using this syntax: npm update myPackage@2.0.0, but I don't know if it would be possible to designate a specific branch. i.e. fix_auth. Is this possible?

**Comment by wrldtvlr** [comment]
Ahh... A little more research suggests I should be able to use this: npm install --g tallyarbiter#fix_auth

**Comment by hrueger** [comment]
> Sorry - my knowledge of npm is very limited

Never mind!

Unfortunately, `npm install --g tallyarbiter#fix_auth` won't work, because this branch is not published to npm. As we're using TypeScript and Angular, some build steps need to happen in order to generate a distributable form of TallyArbiter which is then published to npm.

To run this branch (in development mode), you need to follow the [instructions from the Readme](https://github.com/josephdadams/TallyArbiter#contributing).
After step three you need to switch to this branch using `git checkout fix_auth` and then continue with the other steps

**Comment by wrldtvlr** [comment]
Ok. Uninstalled the npm version, cloned the repo and set the fork. Angular takes forever to compile on a Pi...

**Comment by wrldtvlr** [comment]
Just an update - let things run overnight and not surprisingly found an insufficient memory error when I came in this morning. Increased the memory allocation and restarted things. Will see if that makes any difference. I now remember why I chose the CLI install method previously. Getting the dev environment running on a Pi is not easy. I may need to wait until the auth fix is published before testing.

**Comment by wrldtvlr** [comment]
I wasn't able to get the dev environment for 3.0.2 auth fix installed, so I had to remove it and will reinstall 3.0.1 tomorrow. Will be glad to load it when a package is available.

**Comment by hrueger** [comment]
Never mind, I wouldn't have though that the Angular Compilation is that slow on a pi... May I ask which Pi version do you have? 

Anyway, we'll test this change and release a new version as soon as one of us finds some time to do that ;-)

**Comment by wrldtvlr** [comment]
OK, I'm at a complete loss here. I can't get TA (any version) to recognize the settings or producer password. With the latest CLI package running, the log says this:

2022-03-29 10:36:25] info: User admin (ip addr ::ffff:10.10.12.207) has attempted a login: wrong username or password.
[2022-03-29 10:36:46] info: User producer (ip addr ::ffff:10.10.12.207) has attempted a login: wrong username or password.

The password section of the config.json file in ~/local/share/TallyArbiter says this:

   "username": "producer",
   "password": "$2a$12$RdlK3dTxUkvqpfyXkFPPuuZ4js6YcBBv7Sygxho3340TyGwsp.2Jm",
   "roles": "producer"
  },
  {
   "username": "admin",
   "password": "$2a$12$RdlK3dTxUkvqpfyXkFPPuuZ4js6YcBBv7Sygxho3340TyGwsp.2Jm",
   "roles": "admin"

The hashes were generated at the Bcrypt-generate website for password 12345. We don't even need passwords for TA access. Any suggestions on what might be wrong?

**Comment by hrueger** [comment]
What do you mean with `latest` v2?

**Comment by wrldtvlr** [comment]
Yes. What you get when you do npm install -g tallyarbiter.

**Comment by hrueger** [comment]
This should be 3.0.2
Unfortunately, I don't know what's going on there currently, but I also don't have time to release @MatteoGheza's fix... Sorry for that, its quite busy right now.

**Comment by wrldtvlr** [comment]
I'm at a loss as well. I just tried installing 3.0.1 over 3.0.2 (npm install --global tallyarbiter@3.0.1) as that version was working fine but get the same password error. I assume I'm doing something wrong with the hash, but I can't figure out what.

**Comment by hrueger** [comment]
Are you editing the correct config file? 
We once changed the path to where it's saved, but I don't remember if that was between 1 and 2 or between 2 and 3.

**Comment by wrldtvlr** [comment]
That seems like the most likely. I've been editing ~/.local/share/tallyarbiter/config.json. Is there another location I should look for?

**Comment by wrldtvlr** [comment]
OK - the details are a little fuzzy as I was working on this for several more hours this morning and the steps all blur together, but this is what I think I know as far as installing on a Raspberry Pi:

- Installing TA server on a Raspberry Pi using the developer option (i.e. compiling everything) is almost impossible given the limited memory and cpu.
- Installing using the CLI method is a bit tricky as the permissions are not clear. It seems to be necessary to use sudo npm install rather than just npm install.
- Upgrading from a previous install is very unclear. I think the v2 install location is ~/.local/share/TallyArbiter. The v3.0.2 location when using sudo npm install -global is actually /root/.local/share/TallyArbiter.
- Installing v3.0.2 as an upgrade seems to be missing a security key in the config.json file. Although TA server runs just fine, this results in the password being invalid when attempting to log in.

So, I created a brand new Raspian OS image, reinstalled node and npm (with sudo), then reinstalled TA 3.0.1 using 

sudo npm install -global tallyarbiter@3.0.1.

I then went in and set /root/.local/share/TallyArbiter directory and file permissions to 777 as the changes to the config.json file weren't getting written out.

I'm sure some of this was unnecessary or just plain wrong, but at least everything is running. Once 3.0.2 has been fixed and tested on a Raspberry Pi, I'll try installing that on a clean Raspian image.
